src/main.cpp

```diff
@@ -871,17 +871,29 @@ bool AreInputsStandard(const CTransactionBase& txBase, const CCoinsViewCache& ma
     return true;
 }
 
-unsigned int GetLegacySigOpCount(const CTransactionBase& tx)
+unsigned int GetLegacySigOpCount(const CTransactionBase& txBase)
 {
     unsigned int nSigOps = 0;
-    BOOST_FOREACH(const CTxIn& txin, tx.GetVin())
+    for(const CTxIn& txin: txBase.GetVin())
     {
         nSigOps += txin.scriptSig.GetSigOpCount(false);
     }
-    BOOST_FOREACH(const CTxOut& txout, tx.GetVout())
+    for(const CTxOut& txout: txBase.GetVout())
     {
         nSigOps += txout.scriptPubKey.GetSigOpCount(false);
     }
+
+    if(!txBase.IsCertificate())
+    {
+        const CTransaction& tx = dynamic_cast<const CTransaction&>(txBase);
```

Comment:
check formally that `tx` is not null

Reply:
@alsala In our case we do a cast to reference. If failed, it throws bad_cast exception. Do we need to have any additional check here?

Reply:
We should add a try/catch block, since when dynamic_cast fails for a reference type, it throws a 'bad_cast' exception